python/kvikio/_lib/libnvcomp.pyx

```diff
@@ -57,305 +62,146 @@ class pyNvcompType_t(Enum):
     pyNVCOMP_TYPE_BITS = nvcompType_t.NVCOMP_TYPE_BITS
 
 
-# _Cascaded Compressor / Decompressor
-cdef class _CascadedCompressor:
-    cdef __CascadedCompressor* c
-
-    def __cinit__(self, nvcompType_t t, int num_RLEs, int num_deltas, bool use_bp):
-        self.c = new __CascadedCompressor(
-            t,
-            num_RLEs,
-            num_deltas,
-            use_bp
-        )
+cdef class _nvcompManager:
+    cdef nvcompManagerBase* _impl
```

Comment:
I would rather not. I appreciate your concern here, but the abstraction layer here is pretty sensitive. The complexity here is right on the edge of my comprehension of Cython's capabilities and weaknesses. In particular Cython doesn't support fully abstract virtual classes - anything instantiated in Cython has to have a real vtable. Secondly, Cython doesn't support classes that lack a default constructor. nvcomp C++ is based on the PIMPL pattern, so `nvcompManagerBase` is a purely virtual class, and it also lacks a default constructor.

Attempting to create a stack version of the `nvcompManagerBase` puts Cython into a binary failure mode that is hard to resolve: Either it cannot create a stack instance of `nvcompManagerBase` or it cannot create a concrete instance due to the lack of vtables. I can resolve the first error by dropping `except +` from the child class constructor method, which is not super desirable, but then Cython will fail because it wants to create a concrete instance.

I'm able to use `shared_ptr` with `CompressionConfig` and `DecompressionConfig` because they are not based on the PIMPL pattern. I'm also able to use `shared_ptr` with the last class, `ManagedManager`, because it is actually created purely in C++ inside of the `create_manager` factory function.

I don't doubt that there is some encantation that will make `shared_ptr` usable with `self._impl`, and if you can show me the pattern I'm happy to use it. Otherwise, fiddling with Cython builds could eat up days and not give me any fruit to show, or at best will produce the same result as what I already have.

I'm going to spend a little more time on it today, however, because you pointed out an issue that I created when I worked out the pattern that is used for `ManagedManager` - there used to be a `dealloc: del self._impl` but I dropped it when I created the `ManagedManager` because of a double free. That's not acceptable, so I will work out that fix at least.

My inclination is just to check if `self._mgr` is none, and if so, do the free, otherwise let `shared_ptr` handle it.